### ISSUE TITLE
performance loss in July 2 package update

### ISSUE THREAD

**Issue by jnhwkim** [torch/distro#131]
I experienced performance loss after July 2 package update (2a7637fea5b7f2b6b980b9c5a11c87bd2c861a7d). Here are some results on a project. Anybody has an idea about why this happen?

```
date | hashtag | loss
=====================
Apr 8 | cbb5161 |  6.35515
May 1 | 57c7057  | 6.35515
Jun 22 | a47bec0 |  6.35515
Jul 2 | b64bc52  | 6.35515
Jul 2 | 2a7637f  | *19.6400* <- package update
```

`hashtag` is `torch/distro`'s commit #. `loss` is cross entropy loss at 100 iterations.

**Comment by soumith** [comment]
hi! I am looking into it.

**Comment by soumith** [comment]
do you use cuda? also, do you use cudnn?

**Comment by jnhwkim** [comment]
@soumith use `cudnn`.

**Comment by soumith** [comment]
do you have any ClippedReLU nonlinearity?

**Comment by jnhwkim** [comment]
@soumith I'm sorry for the code is not ready for release. But I can give you any information you need. The model uses Dropout, Tanh, Linear, CMulTable, ParallelTable, CAddTable, optim.rmsprop.

**Comment by soumith** [comment]
okay, this is very helpful.

To convert your model to cudnn, do you use cudnn.convert ?

**Comment by jnhwkim** [comment]
@soumith Oops, I think I did `require 'cudnn'`, however I just called `:cuda()` on the model... May be just `cunn` was used.

**Comment by soumith** [comment]
ok great. i will see what changes were made in cunn at that time. If you run the model in CPU mode (i.e. without GPU), do you also see the loss going to *19.6400* or is it only in GPU mode?

**Comment by jnhwkim** [comment]
@soumith wait a moment, I am checking.

**Comment by soumith** [comment]
Actually, I have found a change in RMSProp that might be affecting you.
Look at this commit: https://github.com/torch/optim/commit/9c08fde975c5998cc25d5ebf265754486dd5c160

Can you see if changing the file:
~/torch/install/share/lua/5.1/optim/rmsprop.lua line 43 to 

"state.m = torch.Tensor():typeAs(x):resizeAs(dfdx):zero()"

and see if that fixes it.

If it does, I can go from there...

**Comment by jnhwkim** [comment]
@soumith Bingo! You're right. I'v compared with `adam` and your idea is confirmed. I am afraid whether this update in `optim.rmsprop` is appropriate because it causes very slow convergence in early iterations. Thanks, tweezers!

**Comment by soumith** [comment]
@jnhwkim in the PR, @andreaskopf suggests that you can now increase the learning rate because the initial vector is properly scaled. Can you try that?

**Comment by soumith** [comment]
* @andreaskoepf

**Comment by jnhwkim** [comment]
@soumith There's no guideline for incresement of the learning rate?

**Comment by andreaskoepf** [comment]
@jnhwkim Could you please try to change the alpha value of rmsProp in your test to `0.9` instead of it's default value of `0.99`. 0.99^100 is still 0.36603 but 0.9^100 is <0,000027. I personally believe an alpha value of 0.99 is too high in most situations (other frameworks also use a lower default value of 0.9, see [tf](https://github.com/tensorflow/tensorflow/blob/bf83048d5f28b7fa0f39df799bd01a8fc581f5cf/tensorflow/g3doc/api_docs/python/functions_and_classes/shard6/tf.train.RMSPropOptimizer.md) or [lasagne](https://github.com/Lasagne/Lasagne/blob/8fe645d28b66f991d547e9b6a314251b8e84446a/lasagne/updates.py#L400)).

I suggested to initialize to 1 instead of zero because otherwise you likely get a very strong amplification of the gradients at the beginning of the training (because you devide by an hypothetical average gradient formed between 0 as the past running average and 0.01 times the actual gradient). To initialize with 1 is conservative and will start close to sgd and only over time normalize with information about the true gradient - how fast this happens depends on the decay parameter alpha.
Regarding learning rates: If I experiment with different optimization schemes I prefer to ideally use the same learnig rate and just vary the optimization methods (e.g. from sgd to rmsprop etc.). With 0 as initialization value and the process often diverged after 3-5 iterations in my experiments.

A solution to make everybody happy could be to simply make the fill value a parameter (and maybe lower alpha to 0.9).

**Comment by jnhwkim** [comment]
@andreaskoepf So, learning rate is the same (I'm starting with 0.0003), but try alpha = 0.9 right?

**Comment by andreaskoepf** [comment]
@jnhwkim The initial learning rate depends on your model and weight initialization and batch size (e.g. how representative was the batch for all examples - for true 1-sample online learning you would probably use a much lower learning rate than for a 200 example batch) - I do not exactly know your setting but you might want to try significantly higher values and see what happens. I have a couple of models that could be started with 1 or 0.1 learning rate with sgd. But for the 'old' rmprop I had to lower the learning rate to something like 1E-4 in order to make it run.

**Comment by jnhwkim** [comment]
@andreaskoepf Thanks for helpful comments! I am bad at tuning.
